### PR TITLE
Add base metric support

### LINKED ISSUE #1290: Add metrics data to RuntimeAppsController's return value
The runtime apps endpoint `runtime/apps` already returns some metric information on a few platforms, for example cpu%,disk%,memory% on PCF.

Enrich the attributes of each AppInstance to contain metrics collected from the Metrics Collector for incoming and outgoing message rates.

### PR DESCRIPTION
**Pull request by markpollack**
Builds on PR https://github.com/spring-cloud/spring-cloud-dataflow/pull/1283

* Removes metric endpoint that took a map of deploymentId->GUID   (this may make the current PR a bit hard to review since there is addition/subtraction noise of a few classes, maybe squash first and diff to last commit on master)
* Metrics are now added to the existing `runtime apps` endpoint in `RuntimeAppsController`

When deploying apps, add the properties
* spring.cloud.stream.metrics.key as `<streamName>.<appLabel>.${spring.cloud.application.guid}`
* `spring.cloud.stream.metrics.properties to` `spring*`

Use AppInstanceStatus's attribute 'guid' to link up to metics collector stats

* Change name and formatting of incoming/outgoing rates to
** `metrics.integration.channel.input.receiveRate`
** `metrics.integration.channel.output.sendRate`
** %.2f formatting

Fixes issue #1290 and replaces PR #1290

You can run using the local deployer/server by first building the local deployer in the PR https://github.com/spring-cloud/spring-cloud-deployer-local/pull/39.

Then build this PR, and run the following commands.

```
in the metrics collector project, create the servers via `./mvnw clean install -PgenerateApps`.  Then cd to `apps/metrics-collector-rabbit` and run

# java -jar target/metrics-collector-rabbit-1.0.0.BUILD-SNAPSHOT.jar --spring.cloud.stream.bindings.input.destination=metrics --spring.cloud.stream.default.contentType=application/json --security.basic.enabled=false --management.security.enabled=false

# java -jar spring-cloud-dataflow-server-local/target/spring-cloud-dataflow-server-local-1.2.0.BUILD-SNAPSHOT.jar --spring.cloud.dataflow.metrics.collector.url=http://localhost:8080/collector/metrics

# shell
dataflow:>app register --name time120RS --type source --uri maven://org.springframework.cloud.stream.app:time-source-rabbit:1.2.0.BUILD-SNAPSHOT

dataflow:>app register --name log120RS --type sink --uri maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.2.0.BUILD-SNAPSHOT

stream create --name foostream --definition "time120RS --spring.cloud.stream.bindings.applicationMetricsChannel.destination=metrics |log120RS --spring.cloud.stream.bindings.applicationMetricsChannel.destination=metrics "

*OR*

stream create --name foostream --definition "time120RS --spring.cloud.stream.bindings.applicationMetrics.destination=metrics |log120RS --spring.cloud.stream.bindings.applicationMetrics.destination=metrics "

if you are using an OOTB built after where https://github.com/spring-cloud/spring-cloud-stream/pull/905 has been merged.

dataflow:>stream deploy --name foostream --properties "deployer.*.count=2,app.*.security.basic.enabled=false,app.*.management.security.enabled=false"

dataflow:> runtime apps
```
Here is sample output for when the count=1

```dataflow:>runtime apps 
╔═════════════════════╤═════════╤══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║App Id / Instance Id │  Unit   │                                                No. of Instances / Attributes                                                 ║
║                     │ Status  │                                                                                                                              ║
╠═════════════════════╪═════════╪══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣
║foostream.log120RS   │deployed │                                                              1                                                               ║
╟┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈╢
║                     │         │                                         guid = 52731                                                                         ║
║                     │         │metrics.integration.channel.input.receiveRate = 1.00                                                                          ║
║                     │         │  metrics.integration.channel.output.sendRate = 0.00                                                                          ║
║                     │         │                                          pid = 17904                                                                         ║
║foostream.log120RS-0 │deployed │                                         port = 52731                                                                         ║
║                     │         │               stderr = /tmp/spring-cloud-dataflow-8493233097930055662/foostream-1491851478581/foostream.log120RS/stderr_0.log║
║                     │         │               stdout = /tmp/spring-cloud-dataflow-8493233097930055662/foostream-1491851478581/foostream.log120RS/stdout_0.log║
║                     │         │                                          url = http://127.0.1.1:52731                                                        ║
║                     │         │                       working.dir = /tmp/spring-cloud-dataflow-8493233097930055662/foostream-1491851478581/foostream.log120RS║
╟─────────────────────┼─────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╢
║foostream.time120RS  │deployed │                                                              1                                                               ║
╟┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈╢
║                     │         │                                         guid = 21432                                                                         ║
║                     │         │metrics.integration.channel.input.receiveRate = 0.00                                                                          ║
║                     │         │  metrics.integration.channel.output.sendRate = 1.00                                                                          ║
║                     │         │                                          pid = 17919                                                                         ║
║foostream.time120RS-0│deployed │                                         port = 21432                                                                         ║
║                     │         │              stderr = /tmp/spring-cloud-dataflow-8493233097930055662/foostream-1491851478746/foostream.time120RS/stderr_0.log║
║                     │         │              stdout = /tmp/spring-cloud-dataflow-8493233097930055662/foostream-1491851478746/foostream.time120RS/stdout_0.log║
║                     │         │                                          url = http://127.0.1.1:21432                                                        ║
║                     │         │                      working.dir = /tmp/spring-cloud-dataflow-8493233097930055662/foostream-1491851478746/foostream.time120RS║
╚═════════════════════╧═════════╧══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝

```

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/MetricStore.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		this.metricsProperties = metricsProperties;
+		restTemplate = new RestTemplate();
+	}
+
+	@HystrixCommand(fallbackMethod = "defaultMetrics")
+	public List<ApplicationsMetrics> getMetrics() {
+		List<ApplicationsMetrics> metrics = null;
+		if (StringUtils.hasText(metricsProperties.getCollector().getUrl())) {
+			try {
+				metrics = restTemplate.exchange(metricsProperties.getCollector().getUrl(), HttpMethod.GET, null,
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/MetricStore.java]: How are Cross Origin Resource (CORS) requests handled in this case (given the REST template is used to connect to external URL from the data flow server)?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/MetricsProperties.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	public Collector getCollector() {
+		return collector;
+	}
+
+	public void setCollector(Collector collector) {
+		this.collector = collector;
+	}
+
+	public static class Collector {
+		private String url;
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/MetricsProperties.java]: Can this be more specific type to denote the URL (just to have some URI scheme validation etc.,)?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java [hunk 3]
```diff
@@ -143,10 +146,39 @@ public RuntimeAppsController(StreamDefinitionRepository streamDefinitionReposito
...
+			}
+		}
+
+		for (AppStatus appStatus : statuses) {
+			Map<String, AppInstanceStatus> appInstanceStatusMap = appStatus.getInstances();
+			appInstanceStatusMap.forEach((k,appInstanceStatus) -> {
+				String trackingKey = appInstanceStatus.getAttributes().get("guid");
+				if (metricsInstanceMap.containsKey(trackingKey)) {
+					ApplicationsMetrics.Instance metricsAppInstance = metricsInstanceMap.get(trackingKey);
+					appInstanceStatus.getAttributes().put("metrics.integration.channel.input.receiveRate",
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java]: The channel name can't always be `input`

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java [hunk 4]
```diff
@@ -250,6 +250,20 @@ private void deployStream(StreamDefinition stream, Map<String, String> streamDep
 			Resource appResource = registration.getResource();
 			Resource metadataResource = registration.getMetadataResource();
 
+			// add properties needed for metrics system
+			appDeployTimeProperties.put("spring.cloud.dataflow.stream.name", currentApp.getStreamName());
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: Can we move these property names to a constant Class (something like `BindingPropertyKeys`)?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/MetricStore.java (hunk 1 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/MetricStore.java]: Mark can probably confirm but currently we don't care as collector have no authentication or authorization so we blindly trust what server asks. So this is simply between scdf server and collector.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java [hunk 5]
```diff
@@ -250,6 +250,20 @@ private void deployStream(StreamDefinition stream, Map<String, String> streamDep
...
+			appDeployTimeProperties.put("spring.cloud.dataflow.stream.app.label", currentApp.getName());
+			appDeployTimeProperties.put("spring.cloud.dataflow.stream.app.type", type.toString());
+			StringBuilder sb = new StringBuilder()
+					.append(currentApp.getStreamName()).append(".")
+					.append(currentApp.getName()).append(".")
+					.append("${spring.cloud.application.guid}");
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: I could have missed the context here. But, can the property `${spring.cloud.application.guid}` always resolve to a specific value? I guess we probably need a value for `default` (some unique number)?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/MetricsProperties.java (hunk 2 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/MetricsProperties.java]: what u mean by more specific type?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java [hunk 6]
```diff
@@ -250,6 +250,20 @@ private void deployStream(StreamDefinition stream, Map<String, String> streamDep
...
+			appDeployTimeProperties.put("spring.cloud.stream.metrics.key", sb.toString());
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: All the `spring.cloud.stream.metrics` properties' keys could probably be defined in `BindingPropertyKeys` Class. Perhaps, we could rename the Class `BindingPropertyKeys` to `SpringCloudStreamPropertyKeys` which would also lead a way to create `SpringCloudDataFlowPropertyKeys` for other property keys that start with prefix `spring.cloud.dataflow`.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java (hunk 5 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: new stuff, required to be set by deployer.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/MetricsProperties.java (hunk 2 above)
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/MetricsProperties.java]: I was thinking about `java.net.URL`. If it is too aggressive, we could at least add some validation for the URL String before it is being used I guess.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java (hunk 5 above)
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: ok, In that case, Can we can refer this from `AppDeployer` constant (once it is added).

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java (hunk 4 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: yeah true, I added some and mark added more. Might be good if we put those to constants somewhere.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java (hunk 3 above)
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java]: Is the `input` supposed to be the `name` of the channel here? If so, the channel name varies based on Spring Cloud Stream configuration for the application.

**Comment by jvalkeal** [comment]
Other than that, this doesn't need no changes to UI as things are directly shown within instance attributes.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java [hunk 7]
```diff
@@ -143,10 +146,39 @@ public RuntimeAppsController(StreamDefinitionRepository streamDefinitionReposito
...
+							String.format("%.2f", metricsAppInstance.getIncomingRate()));
+					appInstanceStatus.getAttributes().put("metrics.integration.channel.output.sendRate",
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java]: Same here on the `output`.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java (hunk 3 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java]: this is just renamed from `incomingRate": 0.9993076938325146` what collector gives. collector itself have this naming issue as `input` is not universal. Issue is tracked there and we need to change things here when collector is made more generic.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/ApplicationsMetrics.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+	public void setApplications(List<Application> applications) {
+		this.applications = applications;
+	}
+
+	public static class Application {
+
+		private String name;
+		private List<Instance> instances;
+		private double incomingRate;
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/ApplicationsMetrics.java]: How does this `incomingRate` different from the `incomingRate` from the `Instance`. Same is for the `outgoingRate`.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/MetricStore.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import com.netflix.hystrix.contrib.javanica.annotation.HystrixCommand;
+
+/**
+ * Store implementation returning metrics info from a collector application.
+ * Implemented via hystrix command having a fallback to empty response.
+ *
+ * @author Janne Valkealahti
+ *
+ */
+@Component
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/MetricStore.java]: Why do we require `@Component` here as the `MetricStore` bean is created from `DataFlowControllerAutoConfiguration`?

Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/support/MetricStore.java]: True, this is a leftover from early development.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/MetricsProperties.java (hunk 2 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/MetricsProperties.java]: We seem to use this format in other places like zuul having
```
		/**
		 * A full physical URL to map to the route. An alternative is to use a service ID
		 * and service discovery to find the physical address.
		 */
		private String url;
```

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java [hunk 10]
```diff
@@ -143,10 +146,39 @@ public RuntimeAppsController(StreamDefinitionRepository streamDefinitionReposito
...
+							String.format("%.2f", metricsAppInstance.getOutgoingRate()));
+				}
+			});
+		}
+
+	}
+
 	@RequestMapping("/{id}")
 	public AppStatusResource display(@PathVariable String id) {
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java]: do you think it is worth adding a REST endpoint in `spring-cloud-dataflow-metrics-collector` that fetches the metrics for a given Stream and incorporate the status with the metrics for that stream here?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java (hunk 5 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: That is actually valid point indeed.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java (hunk 10 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/RuntimeAppsController.java]: collector will have search/filter by stream names, so then it'd be doable.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java (hunk 5 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: created github.com/spring-cloud/spring-cloud-deployer/issues/209

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java (hunk 6 above)
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: created github.com/spring-cloud/spring-cloud-dataflow/issues/1297 to address this.

**Comment by jvalkeal** [comment]
Squashed/merged per https://github.com/spring-cloud/spring-cloud-dataflow/commit/088aed60633c13b80a76a820aa8ccccede14e9d1

**Pull request closed without merging**